Sources/WeaverDI/ThreadSafeBuilderStoringDecorator.swift

```diff
@@ -0,0 +1,47 @@
+//
+//  ThreadSafeBuilderStoringDecorator.swift
+//  WeaverDIPackageDescription
+//
+//  Created by Bruno Mazzo on 5/31/18.
+//
+
+import Foundation
+
+class ThreadSafeBuilderStoringDecorator: BuilderStoring {
+    
+    let builderStoring: BuilderStoring
+    private let queue: DispatchQueue = DispatchQueue(label: "ThreadSafeBuilderStoringDecorator")
+    
+    init(builderStoring: BuilderStoring) {
+        self.builderStoring = builderStoring
+    }
+    
+    func get<B>(for key: InstanceKey, isCalledFromAChild: Bool) -> B? {
+        let semaphore: DispatchSemaphore = DispatchSemaphore(value: 0)
+        var getValue: B? = nil
+        
+        self.queue.async {
+            getValue = self.builderStoring.get(for: key, isCalledFromAChild: isCalledFromAChild)
+            semaphore.signal()
+        }
+        
+        semaphore.wait()
+        return getValue
+    }
+    
+    func set<B>(builder: B, scope: Scope, for key: InstanceKey) {
+        self.queue.async {
```

Comment:
This would need a barrier like so:
```swift
 queue.async(flags: .barrier) { ... } 
```

Reply:
So, the Queue is serial, so the `barrier` flag it's not necessary. If you prefer, I can change the queue to concurrent and then I will need to add the barrier. The difference will be that with a concurrent queue, we will start to have parallels `get` and only the `set`s will be serial. What do you prefer?